cmd/erasure.go

```diff
@@ -18,135 +18,354 @@ package cmd
 
 import (
 	"context"
+	"fmt"
+	"sort"
 	"sync"
+	"time"
 
-	"github.com/klauspost/reedsolomon"
 	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/pkg/bpool"
+	"github.com/minio/minio/pkg/dsync"
+	"github.com/minio/minio/pkg/madmin"
+	"github.com/minio/minio/pkg/sync/errgroup"
 )
 
-// Erasure - erasure encoding details.
-type Erasure struct {
-	encoder                  func() reedsolomon.Encoder
-	dataBlocks, parityBlocks int
-	blockSize                int64
+// OfflineDisk represents an unavailable disk.
+var OfflineDisk StorageAPI // zero value is nil
+
+// partialUpload is a successful upload of an object
+// but not written in all disks (having quorum)
+type partialUpload struct {
+	bucket    string
+	object    string
+	failedSet int
 }
 
-// NewErasure creates a new ErasureStorage.
-func NewErasure(ctx context.Context, dataBlocks, parityBlocks int, blockSize int64) (e Erasure, err error) {
-	e = Erasure{
-		dataBlocks:   dataBlocks,
-		parityBlocks: parityBlocks,
-		blockSize:    blockSize,
-	}
+// erasureObjects - Implements ER object layer.
+type erasureObjects struct {
+	// getDisks returns list of storageAPIs.
+	getDisks func() []StorageAPI
 
-	// Check the parameters for sanity now.
-	if dataBlocks <= 0 || parityBlocks <= 0 {
-		return e, reedsolomon.ErrInvShardNum
-	}
+	// getLockers returns list of remote and local lockers.
+	getLockers func() []dsync.NetLocker
 
-	if dataBlocks+parityBlocks > 256 {
-		return e, reedsolomon.ErrMaxShardNum
-	}
+	// Locker mutex map.
+	nsMutex *nsLockMap
+
+	// Byte pools used for temporary i/o buffers.
+	bp *bpool.BytePoolCap
+
+	// TODO: ListObjects pool management, should be removed in future.
+	listPool *TreeWalkPool
+
+	mrfUploadCh chan partialUpload
+
+	endpoints Endpoints
+}
+
+// NewNSLock - initialize a new namespace RWLocker instance.
+func (er erasureObjects) NewNSLock(ctx context.Context, bucket string, objects ...string) RWLocker {
+	return er.nsMutex.NewNSLock(ctx, er.getLockers, bucket, objects...)
+}
+
+// Shutdown function for object storage interface.
+func (er erasureObjects) Shutdown(ctx context.Context) error {
+	// Add any object layer shutdown activities here.
+	closeStorageDisks(er.getDisks())
+	return nil
+}
+
+// byDiskTotal is a collection satisfying sort.Interface.
+type byDiskTotal []DiskInfo
+
+func (d byDiskTotal) Len() int      { return len(d) }
+func (d byDiskTotal) Swap(i, j int) { d[i], d[j] = d[j], d[i] }
+func (d byDiskTotal) Less(i, j int) bool {
+	return d[i].Total < d[j].Total
+}
 
-	// Encoder when needed.
-	var enc reedsolomon.Encoder
-	var once sync.Once
-	e.encoder = func() reedsolomon.Encoder {
-		once.Do(func() {
-			e, err := reedsolomon.New(dataBlocks, parityBlocks, reedsolomon.WithAutoGoroutines(int(e.ShardSize())))
+// getDisksInfo - fetch disks info across all other storage API.
+func getDisksInfo(disks []StorageAPI, endpoints Endpoints) (disksInfo []DiskInfo, onlineDisks, offlineDisks madmin.BackendDisks) {
```

Comment:
You could have conflicts with this since this PR is merged (https://github.com/minio/minio/commit/9baeda781a60c88875042a15502fb026a57d375b)

Reply:
Fixed